core/store/src/trie/trie_storage.rs

```diff
@@ -467,28 +473,51 @@ impl TrieStorage for TrieCachingStorage {
             None => {
                 self.metrics.shard_cache_misses.inc();
                 near_o11y::io_trace!(count: "shard_cache_miss");
-                // If value is not present in cache, get it from the storage.
-                let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
-                let val = self
-                    .store
-                    .get(DBCol::State, key.as_ref())
-                    .map_err(|_| StorageError::StorageInternalError)?
-                    .ok_or_else(|| {
-                        StorageError::StorageInconsistentState("Trie node missing".to_string())
-                    })?;
-                let val: Arc<[u8]> = val.into();
+                let val;
+                // If data is already being prefetched, wait for that instead of sending a new request.
+                if let Some(prefetcher) = &self.prefetch_api {
+                    let prefetch_state = prefetcher.prefetching.get_or_set_fetching(hash.clone());
+                    // Keep lock until here to avoid race condition between shard cache lookup and reserving prefetch slot.
+                    std::mem::drop(guard);
+
+                    val = match prefetch_state {
+                        PrefetcherResult::SlotReserved | PrefetcherResult::MemoryLimitReached => {
+                            self.read_from_db(hash)?
+                        }
+                        PrefetcherResult::Prefetched(value) => {
+                            near_o11y::io_trace!(count: "prefetch_hit");
+                            value
+                        }
+                        PrefetcherResult::Pending => {
+                            near_o11y::io_trace!(count: "prefetch_pending");
+                            std::thread::yield_now();
+                            // Unwrap: Only main thread  (this one) removes values from staging area,
+                            // therefore blocking read will not return empty.
+                            prefetcher.prefetching.blocking_get(hash.clone()).unwrap()
+                        }
+                    };
+                } else {
+                    // TODO: I think we could drop the guard here. But to preserve previous behavior when prefetching is off in this PR, let's not do that.
+                    val = self.read_from_db(hash)?;
+                }
 
                 // Insert value to shard cache, if its size is small enough.
                 // It is fine to have a size limit for shard cache and **not** have a limit for chunk cache, because key
                 // is always a value hash, so for each key there could be only one value, and it is impossible to have
                 // **different** values for the given key in shard and chunk caches.
                 if val.len() < TrieConfig::max_cached_value_size() {
+                    let mut guard = self.shard_cache.0.lock().expect(POISONED_LOCK_ERR);
```

Comment:
so in case of the "old" behaviour -- we acquire the lock again (while holding it already) - does it lead to deadlocks ?

Reply:
Right, good catch. I think we should just drop the guard in any case above. The only parallel access you could have to the shard cache in the absence of prefetchers would be when multiple chunks for the same shard are applied at the same time. And I see no reason that we need to keep the lock in between, in fact releasing it makes that scenario much more efficient.